[MRG+2] DA, TM, DT repval fix

volrath opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/pydicom/pydicom/blob/master/CONTRIBUTING.md#contributing-pull-requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->

None, afaict.


#### What does this implement/fix? Explain your changes.
<!--
Please summarize the key points of the reference issue, problem or contribution
to facilitate reviewing task. Of course reviewers can always refer to the
original issue but facilitating the reviewing process is much appreciated.
-->

Adds `__repr__` method to `DA`, `TM`, and `DT` values, to make them symmetrical to `IS` and `DS`.

This allows consumers to be able to obtain the original string as the representation of the `DA`/`TM`/`DT` object, instead of something like: `DA(2018, 6, 18)`

#### Any other comments?
<!--
-->

While going through this issue, I also notice that the `DataElement.repval` property tries to use the `original_string` attribute in order to print out the representation string, before falling back to `repr(self.value)`

This, on its own, is fine. However, as far as I could tell, `original_value` is not used/set in `DataElement` instances, but in value instances (`IS`, `DSfloat`, ...) so the `elif` clause in line `pydicom/dataelem.py:307` is never true.

Since I'm not sure I'm missing something here, I just added an extra if clause to the `repval` property, that tries to use the data element value's original string.

If this is indeed a bug, I'm happing to fix it. Also, if checking `DataElement.original_string` is intentional, I can remove the extra if clause that I added and just rely on the value's `__repr__` method.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Thanks for contributing!
-->

darcymason commented:
> This allows consumers to be able to obtain the original string as the representation of the DA/TM/DT object, instead of something like: DA(2018, 6, 18)

Took me a minute to realize this only applies if pydicom.config.datetime_conversion is True. I never use that myself, so I'd be interested in others' opinions as to what the preferred repr should be.  Certainly now it looks odd to have a `DA: DA(...)` in the value text returned from a repr of a dataset.  On the other hand, it is easy to differentiate what you are working with if mucking around on the command line.

> original_value is not used/set in DataElement instances, but in value instances (IS, DSfloat, ...) so the elif clause in line pydicom/dataelem.py:307 is never true.

Yes, I think you are right.  But (as I believe you were saying in your comment's last line) the added check of `self.value.original_string` is not necessary, as the else clause will take care of calling the value's `repr`, which can use `original_string`.

A couple of thoughts going forward:
- it would be nice to have some unit tests for these, to ensure they are working as intended.
- while we are in there, we should change IS to make the quote character consistent - we changed to double quotes in #235,  but there is still the `original_string` clause using single quotes.

volrath commented:
> I never use that myself, so I'd be interested in others' opinions as to what the preferred repr should be.

In my case, I've been using `DataElement.repval` to get the original string of data element _values_ when they're "augmented" with `IS`, `DSfloat`, classes.  That is because until this point, I never saw any documentation reference to `DataElement.value.original_string`, and at the moment I didn't know I had to go to the `DataElement.value` to get the original string.

I see how this might not be the intended use of `repval`.  In that case, I would suggest to fix what I think was the original intent of `DataElement.original_string`, so that it's easily accessable (and findable?) from the DE.

From my limited understanding, this would mean changing `DataElement.value` setter to something like:

```python
    @value.setter
    def value(self, val):
        """Convert (if necessary) and set the `value` of the element."""
        if isString(val):
            self.original_string = val  # <- Notice here.

            # Check if is a string with multiple values separated by '\'
            # If so, turn them into a list of separate strings
            #  Last condition covers 'US or SS' etc
            if self.VR not in \
                ['UT', 'ST', 'LT', 'FL', 'FD', 'AT', 'OB', 'OW', 'OF', 'SL',
                 'SQ', 'SS', 'UL', 'OB/OW', 'OW/OB', 'OB or OW',
                 'OW or OB', 'UN'] and 'US' not in self.VR and _backslash in val:
                val = val.split(_backslash)
        self._value = self._convert_value(val)
```

> Yes, I think you are right. But (as I believe you were saying in your comment's last line) the added check of `self.value.original_string` is not necessary

Correct. Both checks for `original_string` are currently unnecessary.  If we decide to fix/add the original string to `DataElement.original_string` then we can leave `DataElement.repval` unchanged.

> it would be nice to have some unit tests for these, to ensure they are working as intended.

Happy to do so provided your input on how to proceed.

darcymason commented:
> From my limited understanding, this would mean changing DataElement.value setter to something like:

I don't think this is the best approach.  It could work, but it feels like it would be a bit difficult.  For example, with the code you've shown, if someone afterward assigned a new value to the data element that wasn't a string, the `original_string` would still be there from the first assignment.  It makes more sense to me to keep the original string directly with the value that was derived from it.

> > it would be nice to have some unit tests for these, to ensure they are working as intended.
> 
> Happy to do so provided your input on how to proceed.

I think something quite simple should suffice -- perhaps copy the class `DecimalStringtests` in `test_valuerep.py` and replace the code in `setUp` to set datetime_conversion to True, and `tearDown` that sets it False.  Then similarly to the other methods there, just create some instances of the DA, TM and DT and check the repval function returns the correct value.

volrath commented:
Done. Addressed both of your points:


> - it would be nice to have some unit tests for these, to ensure they are working as intended.
> - while we are in there, we should change IS to make the quote character consistent - we changed to double quotes in #235, but there is still the original_string clause using single quotes.

darcymason commented:
@volrath, sorry I let this drop for a while.  Looking again at this, there does seem to still be a failure in the unit tests that needs to be resolved.

volrath commented:
@darcymason no worries.  What do you think we should do with that failing test? it fails cause, as discussed, [I removed the `original_string` check from `DataElem.__repr__`](https://github.com/pydicom/pydicom/pull/660/files#diff-972abd91ef8992fd997cf48c46edad32L307) (since it wasn't being used anywhere in the code). OTOH, I now fear that by removing this attribute, we could break code of other people that are using that attr to set their own stuff?

So our two options are:
1. keep the test, revert d521d7a06aaa2962cabe1dddee60f6fd5ece68e3
2. keep d521d7a06aaa2962cabe1dddee60f6fd5ece68e3, remove the test.

darcymason commented:
Hmm, I think the test should never have happened ... I did a search through the whole code base and I don't think that `<dataelem>.original_string` is used anywhere else (other than the test) in pydicom.  And as I mentioned earlier, it is in danger of getting out of sync with the `<dataelem>.value.original_string`.   I doubt that it has been used in user code, so I think we could just remove it (as the PR already does) and note that in the next release notes.  However, if someone speaks up to the need, we could add a property to redirect it to `<dataelem>.value.original_string` with a deprecation warning.

volrath commented:
@darcymason Couldn't find release notes for the next version here: https://github.com/pydicom/pydicom/tree/master/doc/whatsnew -- I can create one if you want, just need you to tell me what version would be the 'next'.

darcymason commented:
Sure, just create it.  I'd assume 1.2.0 - the minor version increments would generally be for important bug fixes only.

volrath commented:
done @darcymason
